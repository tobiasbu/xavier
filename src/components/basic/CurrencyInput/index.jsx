import React from 'react';
import MaskedInput from 'react-text-mask';
import { conformToMask } from 'text-mask-core';

import * as stringUtils from '@utils/stringUtils';

import NumberInput from '../NumberInput';
import { numberMask } from '../../commons';

/**
 * Click on control button.
 * @param {string} state Input state
 * @param {number} step The state do modify
 */
function onControlClick(state, step) {
  let newValue = stringUtils.currencyToFloat(state) + step;
  if (newValue <= 0) {
    newValue = 0;
  }
  const r = conformToMask(stringUtils.toCurrency(newValue, false), numberMask, state);
  if (!r) {
    return 'R$ 0,00';
  }
  return r.conformedValue || 0;
}

/**
 * Currency input.
 * At the moment this input only accept Reais (R$) currency.
 * This is totally a mess.
 * @param {any} props Currency input props
 */
const CurrencyInput = (props) => {
  const {
    disabled, label, forwardedRef, validation, errorMessage, onChange, defaultValue,
  } = props;

  return (
    <MaskedInput
      mask={numberMask}
      guide
      render={(ref, inputProps) => {
        const maskedProps = inputProps;

        return (
          <NumberInput
            label={label}
            defaultValue={defaultValue}
            step={100}
            disabled={disabled}
            validation={validation}
            errorMessage={errorMessage}
            onControlClick={onControlClick}
            placeholder="R$ 0,00"
            onChange={(e, newVal) => {
              maskedProps.onChange(e);
              if (onChange) {
                onChange(e, newVal);
              }
            }}
            forwardedRef={(e) => {
              ref(e);
              if (forwardedRef) {
                forwardedRef(e);
              }
            }}
            // eslint-disable-next-line react/jsx-props-no-spreading
            onBlur={maskedProps.onBlur}
          />
        );
      }}
    />
  );
};

/**
 * Currency prop types.
 * Same as the number input.
 */
CurrencyInput.propTypes = { ...NumberInput.propTypes };

/**
 * Currency default props.
 */
CurrencyInput.defaultProps = { ...NumberInput.defaultProps };

export default CurrencyInput;
