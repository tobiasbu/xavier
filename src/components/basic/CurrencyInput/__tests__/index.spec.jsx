import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import CurrencyInput from '..';

describe('Currency Input', () => {
  it('should render', () => {
    const component = shallow(<CurrencyInput label="Reais" step={50} />);
    expect(toJSON(component)).toMatchSnapshot();
  });

  // HUM! useRef or React.Ref:
  // https://medium.com/@fabianterh/fixing-typeerror-cannot-read-property-property-of-null-in-react-test-renderer-d91d45137de9

  // it('check currency input value', () => {
  //   const onChange = jest.fn;
  //   const value = 10;
  //   const component = renderer.create(<CurrencyInput label="Dinheiro"
  //  step={50} defaultValue={100} />);
  //   const compJson = component.toJSON();
  //   expect(compJson).toMatchSnapshot();
  //   const input = component.findByType('input');
  // });
});
