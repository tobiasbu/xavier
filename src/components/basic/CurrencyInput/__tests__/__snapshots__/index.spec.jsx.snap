// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Input should render 1`] = `
<t
  guide={true}
  mask={[Function]}
  render={[Function]}
/>
`;
